File: Inverter.vb

Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Drawing
Imports System.Runtime.CompilerServices
Imports System.Threading
Imports System.Windows.Forms

Public Class Inverter
    '  Private Shared __ENCList As New List(Of WeakReference)()
    Private exit_ As Boolean
    Private invertX_ As Boolean
    Private invertY_ As Boolean
    Private pos As Point
    Private running_ As Boolean
    Private swap_ As Boolean


    Public Sub New(x As Boolean, y As Boolean)
        Me.pos = Cursor.Position
        Me.invertX_ = x
        Me.invertY_ = y
        Me.pos = Cursor.Position
    End Sub

    Public Event InvertSettingsChangedEvent(inverter As Inverter, eventArgs As EventArgs)

    Private Sub MouseLoop()
        While Not Me.exit_
            Dim position As Point = Cursor.Position
            Dim x As Integer = 0
            Dim y As Integer = 0
            If Me.swap_ Then
                x = If(Me.InvertX, (Me.pos.X - (position.Y - Me.pos.Y)), (Me.pos.X + (position.Y - Me.pos.Y)))
                y = If(Me.InvertY, (Me.pos.Y - (position.X - Me.pos.X)), (Me.pos.Y + (position.X - Me.pos.X)))
            Else
                x = If(Me.InvertX, (Me.pos.X - (position.X - Me.pos.X)), position.X)
                y = If(Me.InvertY, (Me.pos.Y - (position.Y - Me.pos.Y)), position.Y)
            End If
            If Me.InvertX Then
                If x < 2 Then
                    x = 2
                End If
                If x > (Screen.FromPoint(position).Bounds.Right - 2) Then
                    x = Screen.FromPoint(position).Bounds.Right - 2
                End If
            End If
            If Me.InvertY Then
                If y < 2 Then
                    y = 2
                End If
                If y > (Screen.FromPoint(position).Bounds.Bottom - 2) Then
                    y = Screen.FromPoint(position).Bounds.Bottom - 2
                End If
            End If
            Dim point2 As New Point(x, y)
            Cursor.Position = point2
            Me.pos = Cursor.Position
            Thread.Sleep(5)
        End While
        Me.exit_ = False
    End Sub

    Public Sub Start()
        Me.pos = Cursor.Position
        Me.running_ = True
        Dim thead = New Thread(New ThreadStart(AddressOf Me.MouseLoop)) With { _
          .Priority = ThreadPriority.Highest, _
            .IsBackground = True _
        }
        thead.Start()
    End Sub

    Public Sub [Stop]()
        Me.running_ = False
        Me.exit_ = True
    End Sub

    Public Property InvertX() As Boolean
        Get
            Return Me.invertX_
        End Get
        Set(value As Boolean)
            Me.invertX_ = value
            '     Dim invertSettingsChangedEvent As EventHandler = Me.InvertSettingsChangedEvent
            RaiseEvent InvertSettingsChangedEvent(Me, New EventArgs())
        End Set
    End Property

    Public Property InvertY() As Boolean
        Get
            Return Me.invertY_
        End Get
        Set(value As Boolean)
            Me.invertY_ = value
            '    Dim invertSettingsChangedEvent As EventHandler = Me.InvertSettingsChangedEvent
            RaiseEvent InvertSettingsChangedEvent(Me, New EventArgs())
        End Set
    End Property

    Public ReadOnly Property Running() As Boolean
        Get
            Return Me.running_
        End Get
    End Property
    Public Property Swap() As Boolean
        Get
            Return Me.swap_
        End Get
        Set(value As Boolean)
            Me.swap_ = value
            '   Dim invertSettingsChangedEvent As EventHandler = Me.InvertSettingsChangedEvent
            RaiseEvent InvertSettingsChangedEvent(Me, New EventArgs())
        End Set
    End Property
End Class

File: MouseSwapper/Inverter.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;

using System.Drawing;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Windows.Forms;

public class Inverter
{
	public bool _exit;
	public bool _invertX;
	public bool _invertY;
    public bool _swapXY;
    private Point pos;
	public bool _running;

    public void Set(bool invertX, bool invertY, bool swapXY)
    {
        _invertX = invertX;
        _invertY = invertY;
        _swapXY = swapXY;
    }


	public Inverter()
	{
		pos = Cursor.Position;
	}

	public event InvertSettingsChangedEventEventHandler InvertSettingsChangedEvent;
	public delegate void InvertSettingsChangedEventEventHandler(Inverter inverter, EventArgs eventArgs);

	private void MouseLoop()
	{
		while (!_exit) {
			var position = Cursor.Position;
			int x = 0, y = 0;
			if (_swapXY) {
				x = pos.X + (_invertX ? -1 : 1) * (position.Y - pos.Y);
                y = pos.Y + (_invertY ? -1 : 1) * (position.X - pos.X);
			} else {
                x = _invertX ? (pos.X - (position.X - pos.X)) : position.X;
				y = _invertY ? (pos.Y - (position.Y - pos.Y)) : position.Y;
			}
            var bound = Screen.PrimaryScreen.Bounds;
            bound.Inflate(-4, -4);
            {
				if (x < bound.Left) 
					x = bound.Left;
				else if (x > (bound.Right)) 
					x = bound.Right;
			}
			{
				if (y < bound.Top)
					y = bound.Top;
				else if (y > (bound.Bottom)) 
					y = bound.Bottom;
			}
            pos = Cursor.Position = new Point(x, y);
			Thread.Sleep(5);
		}
		_exit = false;
	}

	public void Start()
	{
		pos = Cursor.Position;
		_running = true;
        _exit = false;
		dynamic thead = new Thread(new ThreadStart(MouseLoop)) {
			Priority = ThreadPriority.Highest,
			IsBackground = true
		};
		thead.Start();
	}

	public void Stop()
	{
		_running = false;
		_exit = true;
	}
}